File: 0011-container-with-most-water/0011-container-with-most-water.py
class Solution:
    def maxArea(self, height: List[int]) -> int:
        #find may area using string
        left, right = 0, len(height) - 1
        max_water = 0
        while left < right:
            max_water = max(max_water, (right-left) * min(height[left], height[right]))

            if height[left] > height[right]:
                right -= 1
            else:
                left +=1
        return max_water

File: 0011-container-with-most-water/0011-container-with-most-water.java
class Solution {
    public int maxArea(int[] height) {
        int maxWater = 0;
        int leftIdx = 0, rightIdx = height.length-1;

        while(leftIdx < rightIdx){
            int width = rightIdx -  leftIdx;
            int current_height = Math.min(height[leftIdx], height[rightIdx]);
            maxWater = Math.max(maxWater, current_height * width);

            while(leftIdx < rightIdx && height[leftIdx] <= current_height){
                leftIdx++;
            }

            while(leftIdx < rightIdx && height[rightIdx] <= current_height){
                rightIdx--;
            }
        }
        return maxWater;
    }
    // public int maxArea(int[] height) {
    //     int maxWater = 0;
    //     int leftIdx = 0, rightIdx = height.length-1;

    //     while(leftIdx < rightIdx){
    //         int width = rightIdx -  leftIdx;
    //         int current_height = Math.min(height[leftIdx], height[rightIdx]);
    //         maxWater = Math.max(maxWater, current_height * width);

    //         if(height[leftIdx] < height[rightIdx]){
    //             leftIdx++;
    //         }else{
    //             rightIdx--;
    //         }
    //     }
    //     return maxWater;
    // }
}